Анализ рекомендаций

nurlantr opened the pull request.

nurlantr committed:
Анализ рекомендаций

ExpertTM commented on src/main/java/school/faang/user_service/messaging/publisher/recommendation/RecommendationEventPublisher.java:
```diff
@@ -0,0 +1,48 @@
+        }
+    }
+
+    public void toEventAndPublish(RecommendationDto recommendationDto) {
```
Зачем тебе нужен этот метод, когда есть у тебя метод publish? Смотрится, как лишним. Ради передачи объекта RecommendationEvent таким способом не лучшая идея. Сделай чз маппер лучше.

nurlantr replied:
да, я все исправил, спасибо

ExpertTM commented on src/main/java/school/faang/user_service/messaging/publisher/recommendation/RecommendationEventPublisher.java:
```diff
@@ -0,0 +1,48 @@
+    private final ObjectMapper objectMapper;
+
+    @Override
+    public void publish(RecommendationEvent event) {
```
Соответственно тут у тебя публичный метод, который нигде не используется и в качестве аргумента принимает объект RecommendationEvent и передачу которого ты реализовал используя метод toEventAndPublish. Лучше сразу сделать логику, где ты будешь вызвать именно метод publish для публикации топика.

nurlantr committed:
Изменения:
- убрал лишний метод в RecommendationEventPublisher и обновил RecommendationService и его тест класс соответственно

ExpertTM commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -71,6 +76,9 @@ public RecommendationDto update(long recommendationId, RecommendationDto updateR
         updatedRecommendation.setSkillOffers(skillOffers);
 
+        RecommendationEvent recommendationEvent = toEvent(recommendationDto);
+        recommendationEventPublisher.publish(recommendationEvent);
```
можно сразу передать toEvent(recommendationDto)

nurlantr replied:
а да, точно! уже исправил

ExpertTM commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -85,4 +93,13 @@ public List<RecommendationDto> getAllUserRecommendations(long recieverId) {
         return recommendationMapper.toListDto(userService.getUserById(authorId).getRecommendationsGiven());
     }
+
+    private RecommendationEvent toEvent(RecommendationDto dto) {
```
Этому методу тут не место. Нарушение SOLID. вынеси лучше это в маппер recommendationMapper

nurlantr replied:
исправил, спасибо)

nurlantr committed:
Изменения

nurlantr committed:
Изменения:
- перенес метод toEntity из RecommendationService в RecommendationMapper и обновил тесты

ExpertTM commented on src/main/java/school/faang/user_service/mapper/recommendation/RecommendationMapper.java:
```diff
@@ -26,4 +25,13 @@ public interface RecommendationMapper {
         return recommendations.stream().map(this::toDto).toList();
     }
+
+    default RecommendationEvent toEvent(RecommendationDto dto) {
```
Подушню еще. Такой способ имеет место быть, но лучше придерживаться одного подхода. Раз у нас тут MapStruct, то лучше использовать его.

egorbarzenkov replied:
согласен, здесь не обязательно делать это через билдер

nurlantr replied:
ахпхапха, окей, исправил

egorbarzenkov commented on src/main/java/school/faang/user_service/event/recommendation/RecommendationEvent.java:
```diff
@@ -0,0 +1,19 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class RecommendationEvent {
+    private long recommendationId;
```
аналогично, в сервисе аналитики добавлено поле private UUID eventId для сквозного идентифицирования, расширь свой евент для поддержания общей логики

nurlantr committed:
Изменения

egorbarzenkov commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -44,9 +43,12 @@ public RecommendationDto create(RecommendationDto recommendationDto) {
         List<SkillOffer> savedSkillOffers = skillOfferService.saveSkillOffers(recommendationDto.getSkillOffers(), savedRecommendation.getId());
         savedRecommendation.setSkillOffers(savedSkillOffers);
 
+        recommendationEventPublisher.publish(recommendationMapper.toEvent(recommendationDto));
```
ты береш данные из recommendationDto, там при создании нет и не должно быть идентификатора рекомендации, он появляется уже после сохранения по этому нужно использовать savedRecommendation, так идентификатор будет

egorbarzenkov commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -71,6 +73,9 @@ public RecommendationDto update(long recommendationId, RecommendationDto updateR
         updatedRecommendation.setSkillOffers(skillOffers);
 
+
+        recommendationEventPublisher.publish(recommendationMapper.toEvent(recommendationDto));
```
аналогично, лучше брать сущность, которая вернулась после сохранения

nurlantr replied:
не уверен что все правильно сделал, но я вроде исправил все недочеты

nurlantr committed:
Изменения

ExpertTM commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -44,9 +45,13 @@ public RecommendationDto create(RecommendationDto recommendationDto) {
+        RecommendationDto savedRecommendationDto = recommendationMapper.toDto(savedRecommendation);
+        recommendationEventPublisher.publish(recommendationEventMapper.toEvent(savedRecommendationDto));
+
         return recommendationMapper.toDto(savedRecommendation);
```
Возвращай сразу savedRecommendationDto. Нет смысла снова из сущности получать Dto

ExpertTM commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -71,6 +76,9 @@ public RecommendationDto update(long recommendationId, RecommendationDto updateR
+        RecommendationDto savedRecommendationDto = recommendationMapper.toDto(updatedRecommendation);
+        recommendationEventPublisher.publish(recommendationEventMapper.toEvent(savedRecommendationDto));
+
         return recommendationMapper.toDto(updatedRecommendation);
```
Аналогично. Возвращай сразу savedRecommendationDto.

nurlantr replied:
понял, исправил

nurlantr committed:
Изменения

egorbarzenkov commented on src/main/java/school/faang/user_service/messaging/publisher/recommendation/RecommendationEventPublisher.java:
```diff
@@ -0,0 +1,36 @@
+            redisTemplate.convertAndSend(recommendationTopic.getTopic(), message);
+        } catch (JsonProcessingException e) {
+            log.error(ExceptionMessages.SERIALIZATION_ERROR + event, e);
+            throw new IllegalArgumentException(ExceptionMessages.SERIALIZATION_ERROR + event, e);
```
смотри в чем может быть проблема, если по какой то причине будет проблема с objectMapper.writeValueAsString или с redisTemplate.convertAndSend, то будут выданы ошибки, которые в новой форме ты будешь пробрасывать дальше по стэку вызова этого метода, ты прервешь основную операцию, и пользователь увидит ошибку вместо созданной или обновленной записи, по которой создание или обновление уже прошло успешно
предлагаю пока оставить только логирование ошибки и убрать throw new IllegalArgumentException

egorbarzenkov commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -44,9 +45,13 @@ public RecommendationDto create(RecommendationDto recommendationDto) {
         savedRecommendation.setSkillOffers(savedSkillOffers);
 
-        return recommendationMapper.toDto(savedRecommendation);
+        RecommendationDto savedRecommendationDto = recommendationMapper.toDto(savedRecommendation);
+        recommendationEventPublisher.publish(recommendationEventMapper.toEvent(savedRecommendationDto));
```
можно так же подумать над тем, чтобы объединить эту логику в отдельный сервис чтобы не городить это здесь)

nurlantr committed:
Изменения

egorbarzenkov commented:
вроде все хорошо, накидай тестов чтобы перепроверить

nurlantr committed:
Добавил тесты
